doc/reference/trd-hil-design.md

```diff
@@ -0,0 +1,713 @@
+which `ErrorCode` is passed.  A common calling pattern is for a trait
+to return `ErrorCode::BUSY` if there is already an operation pending
+and a callback will be issued. This error code is unique in this way:
+the general rule is that `Ok` means there will be a callback in
+response to this call, `Err` with `ErrorCode::BUSY` means this call
+did not start a new operation but there will be a callback in response
+to a prior call, and any other `Err` means there will not be a
+callback.
+
+
+Rule 4: Return Passed Buffers in Error Results
+===============================
+
+Consider this method:
+
+```rust
+fn send(&self, buf: &'static mut [u8]) -> Result<(), ErrorCode>;
+```
+
+This method is for a split-phase call: there is a corresponding
+completion callback that passes the buffer back:
+
+```rust
+fn send_done(&self, buf: &'static mut[u8]);
+```
+
+
+The `send` method follows Rule 2: it returns a synchronous error. But 
+suppose that calling it returns an `Err(ErrorCode)`: what happens to
+the buffer?
+
+Rust's ownership rules mean that the caller can't still hold the reference: 
+it passed the reference to the implementer of `send`. But since the 
+operation did not succeed, the caller does not expect a callback. Forcing
+the callee to issue a callback on a failed operation typically forces it 
+to include an alarm or other timer. Following Rule 1 means it can't do
+so synchronously, so it needs an asynchronous event to invoke the callback
+from. This leads to every implementer of the HIL requiring an alarm or
+timer, which use RAM, has more complex logic, and makes initialization more
+complex.
+
+As a result, in the above interface, if there is an error on `send`, the buffer 
+is lost. It's passed into the callee, but the callee
+has no way to pass it back.
+
+If a split-phase operation takes a reference to a buffer as a parameter, it
+should return a reference to a buffer in the `Err` case:
+
+```rust
+fn send(&self, buf: &'static mut [u8]) -> Result<(), (ErrorCode, &'static mut [u8])>;
+```
+
+Before Tock transitioned to using `Result`, this calling pattern was typically
+implemented with an `Option`:
+
+
+```rust
+fn send(&self, buf: &'static mut [u8]) -> (ReturnCode, Option<&'static mut [u8]>);
+```
+
+In this approach, when the `ReturnCode` is `SUCCESS`, the `Option` is always supposed
+to be `None`; it the `ReturnCode` has an error value, the `Option` contains the passed
+buffer. This invariant, however, cannot be checked. Transitioning to using `Result`
+both makes Tock more in line with standard Rust code and enforces the invariant.
+
+
+Rule 5: Always Pass a Mutable Reference to Buffers
+===============================
+
+Suppose you are designing a trait to write some text to an LCD screen. The trait
+takes a buffer of ASCII characters, which it puts on the LCD:
+
+```rust
+trait LcdTextDisplay {
+  fn display_text(&self, text: &'static [u8]) -> Result<(), ErrorCode>;
```

Comment:
```suggestion
trait LcdTextDisplay {
  // This is an anti-pattern: the `text` buffer should be `mut`, for reasons explained below
  fn display_text(&self, text: &'static [u8]) -> Result<(), ErrorCode>;
```

I think it's useful when including "wrong" code examples to tie a comment directly to the wrong code, to help ensure skimming readers or copy-pasters have locality indicating that this is a negative example.